lib/WWW/Mechanize.pm

```diff
@@ -2028,23 +2064,54 @@ sub submit_form {
         }
     }
 
+    my @filtered_sets;
     if ( $args{with_fields} ) {
         $fields || die q{must submit some 'fields' with with_fields};
-        $self->form_with_fields(keys %{$fields}) or die "There is no form with the requested fields";
+        my @got = $self->all_forms_with_fields(keys %{$fields});
+        die "There is no form with the requested fields" if not @got;
+        push @filtered_sets, \@got;
     }
-    elsif ( my $form_number = $args{form_number} ) {
-        $self->form_number( $form_number ) or die "There is no form numbered $form_number";
+    if ( my $form_number = $args{form_number} ) {
+        my $got = $self->form_number( $form_number );
+        die "There is no form numbered $form_number" if not $got;
+        push @filtered_sets, [ $got ];
     }
-    elsif ( my $form_name = $args{form_name} ) {
-        $self->form_name( $form_name ) or die qq{There is no form named "$form_name"};
+    if ( my $form_name = $args{form_name} ) {
+        my @got = $self->all_forms_with( name => $form_name );
+        die qq{There is no form named "$form_name"} if not @got;
+        push @filtered_sets, \@got;
     }
-    elsif ( my $form_id = $args{form_id} ) {
-        $self->form_id( $form_id ) or die qq{There is no form with ID "$form_id"};
+    if ( my $form_id = $args{form_id} ) {
+        my @got = $self->all_forms_with( id => $form_id );
+        $self->warn(qq{ There is no form with ID "$form_id"}) if not @got;
+        push @filtered_sets, \@got;
     }
-    else {
+
+    if (not @filtered_sets) {
         # No form selector was used.
         # Maybe a form was set separately, or we'll default to the first form.
     }
+    else {
+        # Need to intersect to apply all the various filters.
+        # Assume that each filtered set only has a given form object once.
+        # So we can count occurrences.
+        #
+        tie my %c, 'Tie::RefHash' or die;
```

Comment:
Have you tried writing this code with out `Tie::RefHash`?  If object stringification is not overloaded then it may "just work" without the tied hash.  Would you be able to look into that?